### PR TITLE
Custom NGINX configuration file option

### PR DESCRIPTION
**Pull request by theQuazz**
This will allow people to customize their own NGINX files how they see fit

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -265,3 +268,51 @@ If you use AWS's ECR service to host your images, you may want to consider addin
 If you run `init()` more than once, you'll see errors as wonqa tries to create resources that have already been created by a previous call. You'll need to log into the AWS console to delete these resources and try again.
 
 If the `wonqa.create()` runs but you are unable to accessing your environment causes a timeout (for eg., if it hangs on `Waiting for your QA env to return a 200 OK`), this probably means that the security group that you are using does not have ingress rules for HTTP and HTTPS as well as the rules required by your app to work.
+
+
+## NGINX Config
```
Comment (julesterrien) [inline README.md]: there's an /example dir here which we could use to put this

File: src/create.js [hunk 2]
```diff
@@ -46,7 +47,7 @@ const create = ({
 }) => {
   const scope = {};
   return configure({ WONQA_DIR, awsRegion })
-    .then(() => writeConfFile({ WONQA_DIR, servers }))
+    .then(() => writeConfFile({ WONQA_DIR, servers, configurationPath }))
```
Comment (julesterrien) [inline src/create.js]: have a basic integration test at `create.test.js` which we may want to edit to make sure this value is always passed if it exists

File: src/nginx.js [hunk 3]
```diff
@@ -51,9 +51,17 @@ const buildAndPush = ({
...
-    return `
+const writeConfFile = ({
+  WONQA_DIR,
+  servers,
+  configurationPath,
+}) => new Promise((resolve, reject) => {
+  let config;
+
+  if (configurationPath) {
+    config = fs.readFileSync(configurationPath);
```
Comment (julesterrien) [inline src/nginx.js]: let's add some basic validation on there. maybe something like:
```
const required = [
	“http {”
	“ssl_certificate     /etc/ssl/fullchain1.pem;”
	“ssl_certificate_key /etc/ssl/privkey1.pem;”
	“map $http_upgrade $connection_upgrade {”
    ...all the ones we need
];
required.every(el => config.includes(el));
```

File: src/nginx.js [hunk 4]
```diff
@@ -67,11 +75,10 @@ const writeConfFile = ({ WONQA_DIR, servers }) => new Promise((resolve, reject)
       proxy_set_header    Connection        "upgrade";
     }
   }`;
-  };
 
-  const s = servers.map(({ port, serverName }) => serverConfig(port, serverName));
+    const s = servers.map(({ port, serverName }) => serverConfig(port, serverName));
```
Comment (julesterrien) [inline src/nginx.js]: meant to add the space here?

File: src/nginx.js [hunk 5]
```diff
@@ -67,11 +75,10 @@ const writeConfFile = ({ WONQA_DIR, servers }) => new Promise((resolve, reject)
...
 
-  const config = `
+    config = `
```
Comment (julesterrien) [inline src/nginx.js]: & here

File: src/validate.js [hunk 6]
```diff
@@ -59,7 +60,14 @@ const validateHTTPSOptions = ({
   } else {
     throw new Error('cachePath is undefined');
   }
-  validateNginxConf(servers);
+  if (configurationPath) {
+    const stats = fs.lstatSync(cachePath);
```
Comment (julesterrien) [inline src/validate.js]: meant `configurationPath` here?

File: src/validate.js [hunk 7]
```diff
@@ -59,7 +60,14 @@ const validateHTTPSOptions = ({
...
+    if (!stats.isFile()) {
+      throw new Error('nginx.configurationPath is not a path to a directory');
```
Comment (julesterrien) [inline src/validate.js]: error should say it's not a file

File: src/nginx.js (hunk 4 above)
Comment (theQuazz) [reply src/nginx.js]: Yes, to elevate it to the same level as the block in the if statement

File: src/nginx.js (hunk 5 above)
Comment (theQuazz) [reply src/nginx.js]: Same as other spacing

**Comment by theQuazz** [comment]
@julesterrien updated

File: src/nginx.js [hunk 8]
```diff
@@ -51,9 +51,27 @@ const buildAndPush = ({
...
+}) => new Promise((resolve, reject) => {
+  let config;
+
+  if (configurationPath) {
+    config = fs.readFileSync(configurationPath);
+    const expectedConfigs = [
+      'http {',
+      'ssl_certificate     /etc/ssl/fullchain1.pem;',
+      'ssl_certificate_key /etc/ssl/privkey1.pem;',
+    ];
```
Comment (julesterrien) [inline src/nginx.js]: Let's put all validations in `validate.js` 

Also, should we not validate for others as well? for eg.:
// at least 1 server block
`server {` 

// default server
`listen 443 ssl default_server;` 

// the default server blocks stuff:
`proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;`
`proxy_set_header X-Forwarded-Proto $scheme;`
`proxy_set_header Host $http_host;`
`proxy_redirect off;`
`proxy_http_version 1.1;`
`proxy_set_header    Upgrade           $http_upgrade;`
`proxy_set_header    Connection        "upgrade";`

// events
`events {`
`accept_mutex on;`
`worker_connections 1024;`

File: src/validate.js [hunk 9]
```diff
@@ -59,7 +60,14 @@ const validateHTTPSOptions = ({
   } else {
     throw new Error('cachePath is undefined');
   }
-  validateNginxConf(servers);
+  if (configurationPath) {
```
Comment (julesterrien) [inline src/validate.js]: thoughts on just calling `validateNginxConf({ servers, configurationPath });` here and then doing all nginx validations within that fn?

File: src/nginx.js (hunk 8 above)
Comment (theQuazz) [reply src/nginx.js]: I put it here so we only need to read the file once. And in terms of validation, I'd rather be permissive with what they do in their nginx file, what works for us might not work for them.

**Comment by theQuazz** [comment]
@julesterrien updated

**Comment by julesterrien** [comment]
@theQuazz running into this err when running locally:
screen shot 2018-08-09 at 12 24 14 pm

That's because `ssl.sh` uses the `serverName` values from `nginx.servers` to create the `domains` value used to request SSL certs. To fix, we either need to parse the nginx conf file to find the server names and create the `domains` string (error prone imo) so have the user also provide the `servers` array so we can create this. 
What do you think?

File: src/nginx.js (hunk 8 above)
Comment (julesterrien) [reply src/nginx.js]: ok yea being permissive makes sense to me too. i just want to make sure the nginx conf has at least what it needs to run, for eg. a default server with the basic configs for that server. then again, maybe it's safe to assume that if the user uses a custom config she knows how to set them up

will look into the `events` config - not sure if this is essential or custom to us. do you know?

**Comment by julesterrien** [comment]
sweet approved!

**Pull request merged**